src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/PrivateDataRetrievalService.java

```diff
@@ -1,125 +1,237 @@
+                ApiLogger.info(String.format("Retrieved %d Trusts for Company Number %s", corporateTrustees.getData().size(), hashedTrustId));
+            }
+
+            return corporateTrustees;
+
+        } catch (ApiErrorResponseException e) {
+            if (e.getStatusCode() == HttpServletResponse.SC_NOT_FOUND) {
+                ApiLogger.info("No Corporate Trustee found for Trust Id " + hashedTrustId, logMap);
+                return new PrivateCorporateTrusteeListApi(Collections.emptyList());
+            }
+            throw new ServiceException(e.getStatusMessage(), e);
+        } catch (URIValidationException e) {
+            var message = "Error Retrieving Corporate Trustee data for Trust Id " + hashedTrustId;
+            ApiLogger.errorContext(message, e);
+            throw new ServiceException(e.getMessage(), e);
+        }
+
+    }
+
+    public PrivateTrustDetailsListApi getTrustDetails(String companyNumber)
+            throws ServiceException {
+        var logMap = new HashMap<String, Object>();
+        logMap.put(COMPANY_NUMBER, companyNumber);
+        ApiLogger.info("Retrieving Trusts for Company Number " + companyNumber, logMap);
+
+        try {
+            PrivateTrustDetailsListApi trusts = apiClientService.getInternalApiClient()
+                    .privateTrustDetailsResourceHandler()
+                    .getTrustDetails(OVERSEAS_ENTITY_URI_SECTION + companyNumber + "/trusts/details")
+                    .execute()
+                    .getData();
+
+            if (trusts != null && trusts.getData() != null && !trusts.getData().isEmpty()) {
+                ApiLogger.info(String.format("Retrieved %d Trusts for Company Number %s",
+                        trusts.getData().size(), companyNumber));
+            }
+
+            return trusts;
+        } catch (ApiErrorResponseException e) {
+            if (e.getStatusCode() == HttpServletResponse.SC_NOT_FOUND) {
+                ApiLogger.info("No Trusts found for Company Number " + companyNumber, logMap);
+                return new PrivateTrustDetailsListApi(Collections.emptyList());
+            }
+            throw new ServiceException(e.getStatusMessage(), e);
+        } catch (URIValidationException e) {
+            var message = "Error Retrieving Trust data for " + companyNumber;
+            ApiLogger.errorContext(message, e);
+            throw new ServiceException(e.getMessage(), e);
+        }
+    }
+
+    private String findNonHashedId(String hashedId, String companyNumber) throws ServiceException {
+        PrivateTrustDetailsListApi details = getTrustDetails(companyNumber);
+        Map<String, String> hashedIdMap = getHashedIdMap(details);
+        return hashedIdMap.get(hashedId);
+    }
+
+    private <T extends Hashable> Map<String, String> getHashedIdMap(
+            PrivateDataList<T> hashableDataList) throws ServiceException {
+        Map<String, String> output = new HashMap<>();
```

Comment:
You only need to find the id once. So no need to build a hash map. Suggest you just look once through the list and return the one that matches - if any match.